Source/SwiftLintCore/Models/Baseline.swift

```diff
@@ -0,0 +1,184 @@
+        }
+
+        let violationsByRuleIdentifier = relativePathViolations.groupedByRuleIdentifier(
+            filteredBy: baselineViolations
+        )
+        let baselineViolationsByRuleIdentifier = baselineViolations.groupedByRuleIdentifier(
+            filteredBy: relativePathViolations
+        )
+
+        var filteredViolations: Set<BaselineViolation> = []
+
+        for (ruleIdentifier, ruleViolations) in violationsByRuleIdentifier {
+            guard
+                let baselineViolations = baselineViolationsByRuleIdentifier[ruleIdentifier],
+                    baselineViolations.isNotEmpty else {
+                filteredViolations.formUnion(ruleViolations)
+                continue
+            }
+
+            let groupedRuleViolations = Dictionary(grouping: ruleViolations, by: \.key)
+            let groupedBaselineViolations = Dictionary(grouping: baselineViolations, by: \.key)
+
+            for (key, ruleViolations) in groupedRuleViolations {
+                guard let baselineViolations = groupedBaselineViolations[key] else {
+                    filteredViolations.formUnion(ruleViolations)
+                    continue
+                }
+                if ruleViolations.count > baselineViolations.count {
+                    filteredViolations.formUnion(ruleViolations)
+                }
+            }
+        }
+
+        let violationsWithAbsolutePaths = Set(filteredViolations.violationsWithAbsolutePaths)
+        return violations.filter { violationsWithAbsolutePaths.contains($0) }
+    }
+
+    /// Returns the violations that are present in another `Baseline`, but not in this one.
+    ///
+    /// The violations are filtered using the same algorithm as the `filter` method above.
+    ///
+    /// - parameter otherBaseline: The other `Baseline`.
+    public func compare(_ otherBaseline: Baseline) -> [StyleViolation] {
+        otherBaseline.baselineViolations.flatMap {
+            filter($1.violationsWithAbsolutePaths)
+        }
+    }
+}
+
+private extension Sequence where Element == StyleViolation {
+    var baselineViolations: [BaselineViolation] {
+        var lineCache = LineCache()
+        return map {
+            $0.baselineViolation(text: lineCache.text(at: $0.location))
+        }
+    }
+}
+
+private struct LineCache {
+    private var lines: [String: [String]] = [:]
+
+    mutating func text(at location: Location) -> String {
+        let lineNumber = (location.line ?? 0) - 1
+        guard lineNumber > 0, let absolutePath = location.file else {
+            return ""
+        }
+        if let fileLines = lines[absolutePath] {
+            return (fileLines.isNotEmpty && lineNumber < fileLines.count) ? fileLines[lineNumber] : ""
+        }
+        if let fileLines = SwiftLintFile(path: absolutePath)?.lines.map(\.content) {
+            lines[absolutePath] = fileLines
+            return lineNumber < fileLines.count ? fileLines[lineNumber] : ""
+        }
+        return ""
+    }
+}
```

Comment:
This ought to serve the same purpose but with caching and text extraction not mixed up in a single method:
```suggestion
private struct LineCache {
    private var lines: [String: [String]] = [:]

    mutating func text(at location: Location) -> String {
        let line = (location.line ?? 0) - 1
        if line > 0, let file = location.file, let content = cached(file: file), line < content.count {
            return content[line]
        }
        return ""
    }

    private mutating func cached(file: String) -> [String]? {
        if let fileLines = lines[file] {
            return fileLines
        }
        if let fileLines = SwiftLintFile(path: file)?.lines.map(\.content) {
            lines[file] = fileLines
            return fileLines
        }
        return nil
    }
}
```